openedx/core/djangoapps/course_groups/tests/test_cohorts.py

```diff
@@ -635,6 +636,45 @@ def test_add_user_to_cohort(self, mock_tracker):
             lambda: cohorts.add_user_to_cohort(first_cohort, "non_existent_username")
         )
 
+    @patch("openedx.core.djangoapps.course_groups.cohorts.tracker")
+    def add_user_to_cohorts_race_condition(self, mock_tracker):
+        """
+        Makes use of before_after to force a race condition, in order to
+        confirm handling of such conditions is done correctly.
+        """
+        course_user = UserFactory(username="Username", email="a@b.com")
+        course = modulestore().get_course(self.toy_course_key)
+        CourseEnrollment.enroll(course_user, self.toy_course_key)
+        first_cohort = CohortFactory(course_id=course.id, name="FirstCohort")
+        second_cohort = CohortFactory(course_id=course.id, name="SecondCohort")
+
+        # This before_after contextmanager allows for reliable reproduction of a race condition.
+        # It will break after the first instance of inner_read, and then run add_user_to_cohort again.
+        # Because this second call will write before control is returned, the first call will be writing stale data.
+        # This test confirms that the first add_user_to_cohort call can handle this stale read condition properly.
+        # Proper handling is defined as treating calls as sequential, with write time deciding the order.
+        with before_after.before_after(
+            'django.db.models.Model.save',
```

Comment:
Note that this is the super(...).save() call that will happen when the `CohortMembership.save()` method tries to create as part of `get_or_create()` https://github.com/edx/edx-platform/blob/efischer/cohort_membership_model/openedx/core/djangoapps/course_groups/models.py#L99